File: benchmarks/cross-language/erlang/skynet.erl
%% Erlang Skynet Benchmark
%% Based on https://github.com/atemerev/skynet
%% Uses native Erlang processes (lightweight ~2KB each).
%% For 1M leaves, start with: erl +P 2000000 -noshell -s skynet start
-module(skynet).
-export([start/0]).

get_leaves() ->
    case os:getenv("SKYNET_LEAVES") of
        false ->
            case os:getenv("BENCHMARK_MESSAGES") of
                false -> 1000000;
                Val -> list_to_integer(Val)
            end;
        Val -> list_to_integer(Val)
    end.

%% Units created; also the divisor. See FAIRNESS.md.
-define(SEQ_THRESHOLD, 1000).
total_actors(N) when N =< ?SEQ_THRESHOLD -> 1;
total_actors(N) -> (N div ?SEQ_THRESHOLD) + total_actors(N div 10).

%% Each node spawns 10 children or reports its offset (leaf).
%% Results bubble up via message passing.
%% Same threshold as every other language here.
skynet_node(Offset, Size, Parent) when Size =< ?SEQ_THRESHOLD ->
    Parent ! seq_sum(Offset, Size, 0);
skynet_node(Offset, Size, Parent) ->
    ChildSize = Size div 10,
    Self = self(),
    lists:foreach(fun(I) ->
        spawn(fun() ->
            skynet_node(Offset + I * ChildSize, ChildSize, Self)
        end)
    end, lists:seq(0, 9)),
    Sum = collect(10, 0),
    Parent ! Sum.

seq_sum(_Offset, 0, Acc) -> Acc;
seq_sum(Offset, N, Acc) -> seq_sum(Offset, N - 1, Acc + Offset + N - 1).

collect(0, Acc) -> Acc;
collect(N, Acc) ->
    receive
        Value ->
            collect(N - 1, Acc + Value)
    end.

start() ->
    NumLeaves = get_leaves(),
    TotalActors = total_actors(NumLeaves),

    io:format("=== Erlang Skynet Benchmark ===~n"),
    io:format("Leaves: ~p, concurrency units: ~p (sequential below ~p)~n~n",
              [NumLeaves, TotalActors, ?SEQ_THRESHOLD]),

    Self = self(),
    Start = erlang:monotonic_time(nanosecond),
    spawn(fun() -> skynet_node(0, NumLeaves, Self) end),
    Sum = receive V -> V after 120000 -> timeout end,
    End = erlang:monotonic_time(nanosecond),

    ElapsedNs = End - Start,
    ElapsedUs = ElapsedNs div 1000,

    io:format("Sum: ~p~n", [Sum]),
    if
        ElapsedUs > 0 ->
            NsPerMsg = ElapsedNs div TotalActors,
            ThroughputM = TotalActors div ElapsedUs,
            Leftover = TotalActors - (ThroughputM * ElapsedUs),
            ThroughputFrac = (Leftover * 100) div ElapsedUs,
            io:format("ns/msg:         ~p~n", [NsPerMsg]),
            FracStr = if ThroughputFrac < 10 -> io_lib:format("0~p", [ThroughputFrac]);
                         true -> io_lib:format("~p", [ThroughputFrac])
                      end,
            io:format("Throughput:     ~p.~s M msg/sec~n", [ThroughputM, FracStr]);
        true -> ok
    end,
    halt(0).

File: benchmarks/cross-language/elixir/skynet.exs
#!/usr/bin/env elixir

# Elixir Skynet Benchmark
# Based on https://github.com/atemerev/skynet
# Uses native Erlang/Elixir processes for each tree node (like Erlang implementation).
# For 1M leaves, the BEAM process limit must be raised (pass +P 2000000 to erl).

defmodule Skynet do
  @default_leaves 1_000_000

  defp get_leaves do
    case System.get_env("SKYNET_LEAVES") do
      nil ->
        case System.get_env("BENCHMARK_MESSAGES") do
          nil -> @default_leaves
          val -> String.to_integer(val)
        end
      val -> String.to_integer(val)
    end
  end

  # Total actors = sum of nodes at each level
  # Units created; also the divisor. See FAIRNESS.md.
  @seq_threshold 1000
  defp total_actors(n) when n <= @seq_threshold, do: 1
  defp total_actors(n), do: div(n, @seq_threshold) + total_actors(div(n, 10))

  # Leaf: send offset to parent. Internal: spawn 10 children, collect, sum, report up.
  # Same threshold as every other language here.
  def skynet_node(offset, size, parent) when size <= @seq_threshold do
    send(parent, seq_sum(offset, size, 0))
  end

  def skynet_node(offset, size, parent) do
    child_size = div(size, 10)
    self_pid = self()
    Enum.each(0..9, fn i ->
      spawn(fn -> skynet_node(offset + i * child_size, child_size, self_pid) end)
    end)
    sum = collect(10, 0)
    send(parent, sum)
  end

  defp seq_sum(_offset, 0, acc), do: acc
  defp seq_sum(offset, n, acc), do: seq_sum(offset, n - 1, acc + offset + n - 1)

  defp collect(0, acc), do: acc
  defp collect(n, acc) do
    receive do
      value -> collect(n - 1, acc + value)
    end
  end

  def run do
    num_leaves = get_leaves()
    total = total_actors(num_leaves)

    IO.puts("=== Elixir Skynet Benchmark ===")
    IO.puts("Leaves: #{num_leaves}, concurrency units: #{total} (sequential below #{@seq_threshold})")
    IO.puts("Using Erlang/OTP processes\n")

    caller = self()
    start = :erlang.monotonic_time(:nanosecond)
    spawn(fn -> skynet_node(0, num_leaves, caller) end)
    sum = receive do
      value -> value
    after
      120_000 -> :timeout
    end
    finish = :erlang.monotonic_time(:nanosecond)

    elapsed_ns = finish - start
    elapsed_us = div(elapsed_ns, 1000)

    IO.puts("Sum: #{sum}")
    if elapsed_us > 0 do
      ns_per_msg   = div(elapsed_ns, total)
      throughput_m = div(total, elapsed_us)
      leftover     = total - (throughput_m * elapsed_us)
      frac         = div(leftover * 100, elapsed_us)
      frac_str     = if frac < 10, do: "0#{frac}", else: "#{frac}"
      IO.puts("ns/msg:         #{ns_per_msg}")
      IO.puts("Throughput:     #{throughput_m}.#{frac_str} M msg/sec")
    end
  end
end

Skynet.run()
